### PR TITLE
Add project for  cook book utilities

### PR DESCRIPTION
**Pull request by colombod**
Adding
 - text chunking with overlapping size
 - text chunking by token count

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.AIUtilities/TextChunkingExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+using Microsoft.DeepDev;
+
+namespace Microsoft.DotNet.Interactive.AIUtilities;
+
+public static class TextChunkingExtensions
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities/TextChunkingExtensions.cs:9]: Thoughts on a better name? What else is going to end up in the category of string operations?

Comment (colombod) [reply src/Microsoft.DotNet.Interactive.AIUtilities/TextChunkingExtensions.cs:9]: mm probably will be template parameter parsing?

File: src/Microsoft.DotNet.Interactive.AIUtilities.Tests/SimilaritySearchTests.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        var collection = new List<string>
+        {
+            "diugo",
+            "diago",
+            "carlo",
+            "chuck"
+        };
+
+        Func<string, float[]> stringToFloatArray = a => a.Select(c => (float)c).ToArray();
+        var search = collection.Search(
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities.Tests/SimilaritySearchTests.cs:24]: I don't understand this signature very well. Let's work on simplifying.

File: src/Microsoft.DotNet.Interactive.AIUtilities/SemanticCollection.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+public record ScoredItem<T>(T Item, float Score);
+
+public interface ISimilarityComparer<in T, in TQuery>
+{
+    public float Score(T a, T b);
+    public float Score(T a, TQuery query);
+}
+
+public class CosineSimilarityComparer<T> : ISimilarityComparer<T, float[]>
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities/SemanticCollection.cs:15]: Separate files please.

File: src/Microsoft.DotNet.Interactive.AIUtilities/SemanticCollection.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        return _storage.Remove(item);
+    }
+
+    public int Count => _storage.Count;
+
+    public bool IsReadOnly => false;
+}
+
+public static class IEnumerableSearchExtensions{
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities/SemanticCollection.cs:124]:
```suggestion
public static class EnumerableSearchExtensions{
```

File: src/Microsoft.DotNet.Interactive.AIUtilities.Tests/SimilaritySearchTests.cs [hunk 5]
```diff
@@ -19,10 +19,12 @@ public void can_search_collection_for_similar_element()
             "chuck"
         };
 
-        var search = collection.OrderBySimilarity(
+        var search = collection.ScoreBySimilarity(
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities.Tests/SimilaritySearchTests.cs:22]: Maybe `ScoreBySimilarityTo("diego")`?

**Review by jonsequitur — APPROVED** [review]

**Pull request merged**